proposals/20190909-psp-rules-support.md

```diff
@@ -0,0 +1,50 @@
+# Support for K8s Pod Security Policies (PSPs) in Falco
+
+<!-- toc -->
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  * [Use cases](#use-cases)
+  * [Diagrams](#diagrams)
+  * [Design Details](#design-details)
+
+<!-- tocstop -->
+
+## Summary
+
+We want to make it easier for K8s Cluster Operators to Author Pod Security Policies by providing a way to read a PSP, convert it to a set of falco rules, and then run falco with those rules.
+
+## Motivation
+
+PSPs provide a rich powerful framework to restrict the behavior of pods and apply consistent security policies across a cluster, but it’s difficult to know the gap between what you want your security policy to be and what your cluster is actually doing. Additionally, since PSPs enforce once applied, they might prevent pods from running, and the process of tuning a PSP live on a cluster can be disruptive and painful.
+
+That's where Falco comes in. We want to make it possible for Falco to perform a “dry run” evaluation of a PSP, translating it to Falco rules that observe the behavior of deployed pods and sending alerts for violations, *without* blocking. This helps accelerate the authoring cycle, providing a complete authoring framework for PSPs without deploying straight to the cluster.
+
+### Goals
+
+Transparently read a candidate PSP into an equivalent set of falco rules that can look for the conditions in the PSP.
+
+### Non-Goals
+
+Falco will not automatically read PSPs from a cluster, will not install PSPs, and will not provide guidance on the parts of your infrastructure that are already covered by PSPs. This feature only helps with the testing part of a candidate PSP. For coming up with an initial PSP, you can use tools like [https://github.com/sysdiglabs/kube-psp-advisor](Kube PSP Advisor).
+
+## Proposal
+
+### Use cases
+
+You'll be able to run falco with a `--psp` argument that provides a single PSP Yaml file. Falco will automatcially convert the PSP into an equivalent set of falco rules, load the rules, and then run with the loaded rules. You can optionally provide a `--psp_save=<path>` command line option to save the converted rules to a file.
+
+### Diagrams
+
+No diagrams yet.
+
+### Design Details
+
+* We'll use [inja](https://github.com/pantor/inja) as the templating engine.
```

Comment:
I think this eludes to my question above about what medium we use to create rules. 

In general I think that templating anything is a sign that we need an input `API` with clear methods. In my mind being able to provide tangible input dynamically at runtime seems to be the most effective. 

---

Also what's the dependency plan here?

We are introducing a new dependency into the compiling process, do we want to document this?

Reply:
I'm very scared about this, we need to have a maintainers discussion on wether we want yet another parsing engine in Falco.

Reply:
I'm not worried about adding new third-party code, in the last ~9 months we've added njson, tbb, civetweb, and obviously protobuf/grpc is being added for grpc support. I imagine we'll add prometheus soon so we can collect and export prometheus metrics.

inja isn't really used to parse or interpret anything. It's used to specify a skeleton of what the rules that validate the conditions in a PSP would look like, which gets combined with options/lists/etc to result in a set of rules that validate the conditions in the PSP.

Here's an example of what the template looks like: https://github.com/falcosecurity/falco/blob/psp-rules-support/rules/templates/k8s_psp_rules.yaml.tmpl. I'm not really sure how you can put an API around that, but I'm open to feedback.

Reply:
Ok got your point, thanks for clarifiying

Reply:
to clarify: grpc was already there for CRI, it wasn't added for grpc support